File: bot.py
import datetime
import io
import json
import os
import queue

import requests
import tweepy
from google.cloud import vision
from google.cloud.vision import types

# Load config file
with open('config.json', 'r') as f:
    config = json.load(f)

# Authenticate to Google Vision API
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = config['google']['application-credentials-file']

# Authenticate to Twitter
auth = tweepy.OAuthHandler(config['twitter']['apiKey'], config['twitter']['apiSecret'])
auth.set_access_token(config['twitter']['accessToken'], config['twitter']['accessSecret'])

api = tweepy.API(auth, wait_on_rate_limit=True,
                 wait_on_rate_limit_notify=True)

bannedKeywords = ['image', 'gif', 'giphy', 'tumblr', 'imgur', 'we heart it']


class ResponseTweet:

    def __init__(self, text, in_response):
        self.text = text
        self.in_response = in_response


class TweetStreamListener(tweepy.StreamListener):

    def on_status(self, status):
        if (status.user.id == int(config['twitter']['userId'])
                and 'media' in status.entities and not hasattr(status, 'retweeted_status')):
            image = status.entities['media'][0]['media_url']
            report(annotate(image), image, status.id)


def annotate(path):
    client = vision.ImageAnnotatorClient()

    if path.startswith('http') or path.startswith('gs:'):
        image = types.Image()
        image.source.image_uri = path

    else:
        with io.open(path, 'rb') as image_file:
            content = image_file.read()

        image = types.Image(content=content)

    web_detection = client.web_detection(image=image).web_detection

    return web_detection


def report(annotations, image_url, tweet_id):
    global counter, id_length
    if counter % 2 == 0:
        pages = []
        matches = []
        partial_matches = []
        keywords = {}
        include_url = False
        if annotations.pages_with_matching_images:
            for page in annotations.pages_with_matching_images:
                pages.append(page.url)

        if annotations.full_matching_images:
            for image in annotations.full_matching_images:
                matches.append(image.url)

        if annotations.partial_matching_images:
            for image in annotations.partial_matching_images:
                partial_matches.append(image.url)

        if annotations.web_entities:
            max_score = annotations.web_entities[0].score
            for entity in annotations.web_entities:
                if len(entity.description) > 0:
                    keywords[entity.description] = entity.score
            limit = 230 if include_url else 265 - id_length
            tweet = '. @archillect Related keywords: "'
            for entity in annotations.web_entities:
                keyword = entity.description
                if len(tweet) + len(keyword) <= limit:
                    if len(keyword) > 0 and keyword.lower() not in bannedKeywords:
                        tweet += keyword + ', '
                else:
                    break
            if max_score > 0.35:
                data = {
                    'image': image_url,
                    'pages': pages,
                    'matches': matches,
                    'partialMatches': partial_matches,
                    'keywords': keywords,
                    'archillectTweet': int(tweet_id),
                    'tweet': 0,
                    'key': config['contextMonster']['apiKey']
                }
                r = requests.post(url=config['contextMonster']['apiUrl'], json=data)
                if r.status_code == requests.codes.created:
                    tweet = tweet[:-2]
                    monster_id = json.loads(r.text)['id']
                    if include_url:
                        tweet += '" Full report: ' + config['contextMonster']['reportUrl'] + str(monster_id)
                    else:
                        tweet += '" Ref: ' + str(monster_id)

                    if q.qsize() == 3:
                        while not q.empty():
                            enqueued = q.get()
                            api.update_status(enqueued.text, in_reply_to_status_id=enqueued.in_response)
                            print(datetime.datetime.now().strftime("%H:%M:%S") + " Published: " + enqueued.text)
                    q.put(ResponseTweet(tweet, tweet_id))
                    print(datetime.datetime.now().strftime("%H:%M:%S")
                          + " Added to queue (" + str(q.qsize()) + "): " + tweet)
                    id_length = len(str(monster_id + 1))
    counter += 1


counter = 0
id_length = 3
q = queue.Queue()
tweetStreamListener = TweetStreamListener()
myStream = tweepy.Stream(auth=api.auth, listener=tweetStreamListener)
myStream.filter(follow=[config['twitter']['userId']])

File: bot.js
const Twit = require("twit");
const vision = require("@google-cloud/vision");
const database = require("./tweets_db");
const alertSender = require("./alertmailer");
const config = require("./config.json");
process.env["GOOGLE_APPLICATION_CREDENTIALS"] =
    config.google_vision_api.application_credentials_file;

const twit = new Twit(config.twitter);
const visionClient = new vision.ImageAnnotatorClient();
const archillectID = config.twitter.archillect_id;
const stream = twit.stream("statuses/filter", { follow: archillectID });
const responsesQueue = [];
let archillectTweets = 0;

sendStartupNotification();
streamTweets();

function sendStartupNotification() {
    alertSender.mail("Bot started", "Archillect Context started");
}

function streamTweets() {
    stream.on("tweet", tweet => {
        handleArchillectTweet(tweet);
    });
}

async function handleArchillectTweet(tweet) {
    if (!isValidArchillectTweet(tweet)) return;
    if (archillectTweets++ % 2 != 0) return;
    const image = tweet.entities.media[0].media_url_https;
    const visionResult = await getVisionResult(image);
    const keywords = getRelatedKeywords(visionResult);
    if (keywords.length === 0) return;
    const response = `.@archillect Related keywords: "${keywords.join(", ")}"`;
    handleResponse(tweet.id_str, response);
    database.insert(tweet, visionResult);
}

function isValidArchillectTweet(tweet) {
    return (
        tweet.user.id === archillectID &&
        !tweet.hasOwnProperty("retweeted_status") &&
        tweet.entities.media.length === 1
    );
}

async function getVisionResult(image) {
    const [result] = await visionClient.webDetection(image);
    return result.webDetection;
}

function getRelatedKeywords(visionResult) {
    const keywords = [];
    for (const label of visionResult.webEntities) {
        const keyword = label.description;
        if (!isValidKeyword(keyword)) continue;
        keywords.push(keyword);
    }
    return keywords;
}

function isValidKeyword(keyword) {
    const bannedKeywords = config.bot.banned_keywords;
    const urlRegex = new RegExp(config.bot.url_regex);
    return (
        keyword !== "" &&
        !bannedKeywords.includes(keyword.toLowerCase()) &&
        !urlRegex.test(keyword)
    );
}

function handleResponse(archillectTweetId, response) {
    const responsesQueueMaxLength = config.bot.responses_queue_max_length;
    responsesQueue.push({ tweet_id: archillectTweetId, response: response });
    console.log("Added to queue: " + response);
    if (responsesQueue.length === responsesQueueMaxLength) {
        postAllResponses();
    }
}

function postAllResponses() {
    while (responsesQueue.length > 0) {
        const response = responsesQueue.shift();
        const params = {
            status: response.response,
            in_reply_to_status_id: response.tweet_id
        };
        twit.post("statuses/update", params, responseCallback);
    }
}

async function responseCallback(error, data, _r) {
    if (error) {
        console.log(error);
        await alertSender.mail("Twitter API error", error);
        process.exit(1);
    } else {
        console.log("Published: " + data.text);
    }
}
